File: Sorting/LNT_no532_Reverse Pairs.py
class Solution:
    """
    @param A: an array
    @return: total of reverse pairs
    """

    def reversePairs(self, A):
        if len(A) < 2:
            return 0

        self.temp = [0 for _ in range(len(A))]
        self.count = 0
        self.merge_sorter(A, 0, len(A) - 1)
        return self.count

    def merge_sorter(self, A, left, right):
        if left >= right:
            return

        mid = (left + right) // 2
        self.merge_sorter(A, left, mid)
        self.merge_sorter(A, mid + 1, right)
        self.merger(A, left, right)

    def merger(self, A, left, right):

        mid = (left + right) // 2
        i, j = left, mid + 1
        index = left

        while i <= mid and j <= right:
            if A[i] > A[j]:
                self.temp[index] = A[j]
                j += 1
                self.count += mid - i + 1
            else:
                self.temp[index] = A[i]
                i += 1
            index += 1

        while i <= mid:
            self.temp[index] = A[i]
            i += 1
            index += 1
        while j <= right:
            self.temp[index] = A[j]
            j += 1
            index += 1

        for index2 in range(left, right + 1):
            A[index2] = self.temp[index2]

File: Sorting/LNT_no532_Reverse Pairs.cpp
class Solution {
public:
    /**
     * @param A: an array
     * @return: total of reverse pairs
     */
    long long reversePairs(vector<int> &nums) {
        int n = nums.size();
        if (n < 2)
            return 0;
        
        vector<int> temp(n, 0);
        return mergeSort(nums, temp, 0, n - 1);
    }

    long long mergeSort(vector<int>& nums, vector<int>& temp, int left, int right) {
        if (left >= right)
            return 0;
        
        int mid = (left + right) / 2;
        long long res = 0;
        res += mergeSort(nums, temp, left, mid);
        res += mergeSort(nums, temp, mid + 1, right);
        res += merger(nums, temp, left, right);
        
        return res;
    }
    
    long long merger(vector<int>& nums, vector<int>& temp, int left, int right) {
        int mid = (left + right) / 2;
        int i = left, j = mid + 1, index = left;
        long long res = 0;
        
        while (i <= mid && j <= right) {
            if (nums[i] > nums[j]) {
                temp[index] = nums[j];
                j++;
                res += mid - i + 1;
            }
            else {
                temp[index] = nums[i];
                i++;
            }
            index++;
        }
        while (i <= mid) {
            temp[index] = nums[i];
            i++;
            index++;
        }
        while (j <= right) {
            temp[index] = nums[j];
            j++;
            index++;
        }
        
        for (index = left; index <= right; index++)
            nums[index] = temp[index];
        
        return res;
    }
};